service/matching/version_sets.go

```diff
@@ -248,3 +248,50 @@ func makeVersionInSetDefault(data *persistence.VersioningData, setIx, versionIx
 		setVersions[len(setVersions)-1] = moveMe
 	}
 }
+
+func lookupVersionSetForPoll(data *persistencespb.VersioningData, caps *commonpb.WorkerVersionCapabilities) (string, error) {
+	// for poll, only the latest version in the compatible set can get tasks
+	// find the version set that this worker is in
+	setIdx, _ := findVersion(data, caps.BuildId)
+	if setIdx < 0 {
+		return "", errBuildNotFound
+	}
+	set := data.VersionSets[setIdx]
+	if caps.BuildId != set.BuildIds[len(set.BuildIds)-1].Id {
+		return "", errNewerBuildFound
+	}
+	return getSetID(set), nil
```

Comment:
Can be multiple set IDs, how are you handling this?
Next PR?

Reply:
I thought you'd return all set IDs and pick one at random when matching with a task.

Reply:
I get that there needs to be a single point where the poller and task meet.
For sync match (add `Add*Task`) APIs we'll want to use a route to a single set ID and make sure we spool to the same task queue storage, but when we poll from persistence we'll need to poll from all set IDs.
Polling from storage can happen concurrently from both set IDs, I was thinking we choose a set ID at random and if there are no tasks, continue to the next set ID.

I haven't given it that much thought though and you probably have more context there.

Reply:
I think you're thinking of things happening at a slightly different layer than I'm planning.

> when we poll from persistence we'll need to poll from all set IDs

not exactly: when we poll we need to poll from the logical _set_, not any specific ids. we pick one id arbitrarily to represent the set (so that I don't have to do a giant refactoring of everything)

> Polling from storage can happen concurrently from both set IDs

call it "reading" from storage since our construct is a "taskReader". and yes, it happens concurrently for all ids. but then those tasks will get redirected to the tqm for one of those ids (arbitrarily, as long as it matches the one we use for poll)

admittedly this is not how it would look if we designed matching from scratch with these requirements. but it is a reasonable structure that should work

Reply:
Just keep in mind that you'll need to extract all of the set IDs for constructing the multiple readers.
That's fine if it happens at a different layer as long as it happens (and has test coverage).